File: Introduction_to_Robotics_Course/Chapter1/Lesson1/Chapter1_Lesson1.cpp

#include <iostream>
#include <Eigen/Dense>

struct MinimalRobot {
    double x_r; // 1D internal state
    explicit MinimalRobot(double x0=0.0) : x_r(x0) {}

    double sense(double x_e) const {
        return x_e - x_r; // y = h(x_r, x_e)
    }

    double policy(double y, double alpha=0.0, double r=0.0) const {
        double phi = 0.5 * y;
        return alpha * r + (1.0 - alpha) * phi; // u(t)
    }

    void actuate(double u, double dt=0.1) {
        x_r += u * dt; // x_r_dot = u
    }
};

int main() {
    MinimalRobot robot(0.0);
    double x_e = 10.0;

    for(int k=0; k<50; ++k){
        double y = robot.sense(x_e);
        double u = robot.policy(y, 0.2, 0.0);
        robot.actuate(u);
    }
    std::cout << "Final robot position: " << robot.x_r << std::endl;
    return 0;
}
      

File: Introduction_to_Robotics_Course/Chapter1/Lesson1/Chapter1_Lesson1.java

public class MinimalRobot {
    private double x_r;

    public MinimalRobot(double x0){
        this.x_r = x0;
    }

    public double sense(double x_e){
        return x_e - x_r;
    }

    public double policy(double y, double alpha, double r){
        double phi = 0.5 * y;
        return alpha * r + (1.0 - alpha) * phi;
    }

    public void actuate(double u, double dt){
        x_r += u * dt;
    }

    public static void main(String[] args){
        MinimalRobot robot = new MinimalRobot(0.0);
        double x_e = 10.0;

        for(int k=0; k<50; k++){
            double y = robot.sense(x_e);
            double u = robot.policy(y, 0.2, 0.0);
            robot.actuate(u, 0.1);
        }
        System.out.println("Final robot position: " + robot.x_r);
    }
}
      